### PR TITLE
Generic for overlapping types

### LINKED ISSUE #370: Support Generic derivation for overlapping types
Given the following ADT
```
sealed trait Root

sealed trait A extends Root
case class A1(s: String) extends A

sealed trait B extends Root
case class B1(s: String) extends B

case class AB(s: String) extends A with B
```
`Generic[Root]` fails to compile in 2.1.0 and 2.2.0-RC4.

It would be nice to have
`Generic[Root]{type Repr = A1 :+: B1 :+: AB :+: CNil}` 
as a result

### PR DESCRIPTION
**Pull request by notxcain**
This PR resolves #370.

### CODE REVIEW

**Comment by milessabin** [comment]
Many thanks ... much appreciated! Merging :-)

**Comment by notxcain** [comment]
:+1:

File: core/src/test/scala/shapeless/generic.scala [hunk 1]
```diff
@@ -265,6 +272,17 @@ class GenericTests {
     typed[AbstractSingle](s1)
   }
 
+  @Test
+  def testOverlappingCoproducts {
+    val gen = Generic[Overlapping]
+    val o: Overlapping = OAB(1)
+    val o0 = gen.to(o)
+    typed[OAB :+: OAC :+: OBC :+: CNil](o0)
```
Comment (OlivierBlanvillain) [inline core/src/test/scala/shapeless/generic.scala]: I think it's a bit weird to auto-flatten the type hierarchy like this, wouldn't it make sense to infer `type Repr = OA :+: OB :+: CNil` instead?

Comment (milessabin) [reply core/src/test/scala/shapeless/generic.scala]: I could see a case being made for that.

**Pull request merged**